packages/calcite-components/src/tests/commonTests.ts

```diff
@@ -1839,3 +1841,133 @@ export function openClose(componentTagOrHTML: TagOrHTML, options?: OpenCloseOpti
     });
   }
 }
+
+/**
+ *
+ * Helper to test custom theming of a component's associated tokens.
+ *
+ * @example
+ * describe("default", () => {
+ *   const tokens = {
+ *     "--calcite-action-bar-trigger-background-color": {
+ *       selector: "calcite-action-bar",
+ *       shadowSelector: "calcite-action-group calcite-action >>> .button",
+ *       targetProp: "backgroundColor",
+ *     },
+ *     "--calcite-action-bar-trigger-background-color-active": {
+ *       selector: "calcite-action-bar",
+ *       shadowSelector: "calcite-action-group calcite-action >>> .button",
+ *       targetProp: "backgroundColor",
+ *       state: { press: { attribute: "class", value: CSS.expandToggle } },
+ *     },
+ *     "--calcite-action-bar-trigger-background-color-focus": {
+ *       selector: "calcite-action-bar",
+ *       shadowSelector: "calcite-action-group calcite-action >>> .button",
+ *       targetProp: "backgroundColor",
+ *       state: "focus",
+ *     },
+ *     "--calcite-action-bar-trigger-background-color-hover": {
+ *       selector: "calcite-action-bar",
+ *       shadowSelector: "calcite-action-group calcite-action >>> .button",
+ *       targetProp: "backgroundColor",
+ *       state: "hover",
+ *     },
+ *   };
+ *   themed(`calcite-action-bar`, tokens);
+ * });
+ *
+ * @param componentTagOrHTML  - The component tag or HTML markup to test against.
+ * @param tokens - A record of token names and their associated selectors, shadow selectors, target props, and states.
+ */
+export function themed(
+  componentTagOrHTML: TagOrHTML,
+  tokens: Record<
+    string,
+    {
+      selector: string;
+      shadowSelector?: string;
+      targetProp: string;
+      state?: string | Record<string, { attribute: string; value: string }>;
+    }
+  >,
+): void {
+  it("is theme-able", async () => {
+    const page = await simplePageSetup(componentTagOrHTML);
+    const expectChecklist: Record<
+      string,
+      [
+        E2EElement,
+        Record<
+          string,
+          [
+            E2EElement,
+            Record<string, [string, string, string | Record<string, { attribute: string; value: string }> | undefined]>,
+          ]
+        >,
+      ]
+    > = {};
+
+    for (const token in tokens) {
+      const { selector, shadowSelector, targetProp, state } = tokens[token];
+      const themedTokenValue = assignTestTokenThemeValues(token);
+      if (!expectChecklist[selector]) {
+        const el = await page.find(selector);
+        expectChecklist[selector] = [el, {}];
+      }
+
+      const [el, selectors] = expectChecklist[selector];
+
+      if (!selectors[shadowSelector || selector]) {
+        let target: E2EElement;
+
+        if (shadowSelector && shadowSelector.includes(">>>")) {
+          const shadowSelectors = shadowSelector.split(" ");
+
+          for (let i = 0; i < shadowSelectors.length; i++) {
+            const s = shadowSelectors[i];
+
+            if (i === 0) {
```

Comment:
I'm not sure I follow this. Could you shed some light on how this works? 

Also, has custom deep piercing been working smoothly? I recall this wasn't working before due to a [Stencil bug](https://github.com/ionic-team/stencil/pull/5481) that was recently fixed.

Reply:
This uses the Stencil bug work around that you posted. It does expect a certain pattern to the shadowSelector but it does work.

`[shadowSelector] [wrapper-class] [sub-element] >>> [wrapper-class] [sub-element] >>> [shadowSelector]`